File: 25.cpp
#include<iostream>
#include "ListNode.h"

using namespace std;

class Solution {
public:
    ListNode* reverseKGroup(ListNode* head, int k) {
        int cnt=0;
        ListNode preHead(-1);
        ListNode* prev=&preHead;
        prev->next=head;
        while(head){
            cnt++;
            head=head->next;
        }
        while(cnt>=k){
            ListNode* cur=prev->next;
            ListNode* nex=cur->next;
            for(int i=1;i<k;++i){
                cur->next=nex->next;
                nex->next=prev->next;
                prev->next=nex;
                nex=cur->next;
            }
            prev=cur;
            cnt-=k;
        }
        return preHead.next;
    }
};

int main(){
	ListNodeEnv l;
	vector<int> v={1,2,3,4,5,6,7,8,9,10};
	ListNode* head=l.initialize(v);
	l.showList(head);
	int k=3;

	Solution s;
	ListNode* newList=s.reverseKGroup(head,k);
	l.showList(newList);

	return 0;
}

File: Python/25.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def reverseKGroup(self, head, k):
        """
        :type head: ListNode
        :type k: int
        :rtype: ListNode
        """
        if head == None:
            return head
        prevHead = prev = ListNode(-1)
        next = head
        i=0
        while True:
            while i < k and next != None:
                next = next.next
                i+=1
            if i == k:
                prev.next = self.reverseList(head, next)
                prev = head
                head = next
                i = 0
            else:
                prev.next = head
                break
        return prevHead.next  
        
    def reverseList(self, head, end):
        prev = None
        while head != end:
            next = head.next
            head.next = prev
            prev = head
            head = next
        return prev
